packages/lib/src/vis/scatter/ScatterPoints.tsx

```diff
@@ -71,28 +72,46 @@ function ScatterPoints(props: Props) {
     [onPointerOut]
   );
 
-  const [dataGeometry] = useState(() => new BufferGeometry());
+  const { length } = data;
+  const { abscissaScale, ordinateScale } = useVisCanvasContext();
   const invalidate = useThree((state) => state.invalidate);
 
-  const dataToColorScale = useDataToColorScale(
+  const indexToPosition = useIndexToPosition(
+    abscissas,
+    abscissaScale,
+    ordinates,
+    ordinateScale
+  );
+
+  const valueToColor = useValueToColor(
     scaleType,
     domain,
     colorMap,
     invertColorMap
   );
 
-  const { position, color } = useBufferAttributes(
-    abscissas,
-    ordinates,
-    data,
-    dataToColorScale
-  );
+  const dataGeometry = useMemo(() => {
+    const geometry = new BufferGeometry();
+    geometry.setAttribute('position', createBufferAttr(length, 3));
+    geometry.setAttribute('color', createBufferAttr(length, 3));
+    return geometry;
+  }, [length]);
 
   useLayoutEffect(() => {
-    dataGeometry.setAttribute('position', new BufferAttribute(position, 3));
-    dataGeometry.setAttribute('color', new BufferAttribute(color, 3, true));
+    const { color, position } = dataGeometry.attributes;
+
+    data.forEach((val, index) => {
+      const { x, y, z } = indexToPosition(index);
+      position.setXYZ(index, x, y, z);
+
+      const { r, g, b } = valueToColor(val);
+      color.setXYZ(index, r / 255, g / 255, b / 255); // normalize RGB channels
```

Comment:
1. Here, I'm doing the normalisation myself to avoid the `Uint8Array` to `Float32Array` conversion, and also to be more explicit (the "normalized" feature is enabled by passing `true` as third argument to `new BufferAttribute()`, which I find very obscure).
2. `valueToColor` now returns an `RgbColor` directly instead of an array of three numbers. This saves us a spread or a loop through those three values (when calling `BufferAttribute#set`).